physicalplan/scan/literal_selector.go

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) The Thanos Community Authors.
+// Licensed under the Apache License 2.0.
+
+package scan
+
+import (
+	"context"
+	"math"
+	"time"
+
+	"github.com/thanos-community/promql-engine/physicalplan/model"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql"
+)
+
+// numberLiteralSelector returns []model.StepVector with same sample value across time range.
+type numberLiteralSelector struct {
+	vectorPool *model.VectorPool
+
+	mint        int64
+	maxt        int64
+	step        int64
+	currentStep int64
+	stepsBatch  int
+
+	val  float64
+	call FunctionCall
+}
+
+func NewNumberLiteralSelector(pool *model.VectorPool, mint, maxt time.Time, step time.Duration, stepsBatch int, val float64, call FunctionCall) model.VectorOperator {
+	return &numberLiteralSelector{
+		vectorPool:  pool,
+		mint:        mint.UnixMilli(),
+		maxt:        maxt.UnixMilli(),
+		step:        step.Milliseconds(),
+		currentStep: mint.UnixMilli(),
+		stepsBatch:  stepsBatch,
+		val:         val,
+		call:        call,
+	}
+}
+
+func (o *numberLiteralSelector) Series(ctx context.Context) ([]labels.Labels, error) {
+	// If number literal is included within function, []labels.labels must be initialized.
+	if o.call != nil {
+		return []labels.Labels{labels.New()}, nil
+	}
+	return make([]labels.Labels, 1), nil
+}
+
+func (o *numberLiteralSelector) GetPool() *model.VectorPool {
+	return o.vectorPool
+}
+
+func (o *numberLiteralSelector) Next(ctx context.Context) ([]model.StepVector, error) {
+	if o.currentStep > o.maxt {
+		return nil, nil
+	}
+
+	totalSteps := int64(1)
+	if o.step != 0 {
```

Comment:
I wonder if we can extract this logic somewhere so we don't have to repeat it. My initial idea was to add a `NumSteps()` method to the `query.Options`, and inject an instance in relevant scanners.

Reply:
Yup! That seems good to me too, but seems like we don't use `query.Options` anywhere else. Let me try to refactor this a bit in the next PR! 🙂